kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs

```diff
@@ -279,6 +282,385 @@ impl<'tcx> GotocCtx<'tcx> {
+                        .filter_map(|idx| {
+                            let o = &operands[idx];
+                            let oty = self.operand_ty(o);
+                            let o_layout = self.layout_of(oty);
+                            if o_layout.size.bytes() == 0 {
+                                None
+                            } else {
+                                Some(self.codegen_operand(o))
+                            }
+                        })
+                        .collect();
+                    let mut operands_base_offset: u64 = 0;
+                    if let FieldsShape::Arbitrary { offsets, .. } = &variant_layout.fields {
+                        let mut indices = variant_layout.fields.index_by_increasing_offset();
+                        if let Some(idx) = indices.next() {
+                            operands_base_offset = offsets[idx].bits();
+                        }
+                    } else {
+                        unreachable!();
+                    }
+                    let mut starting_offset: u64 = 0;
+                    let untagged_member_expr = self.codegen_niche_struct_from_operands_rec(
+                        &mut variant_operands.iter().peekable(),
+                        operands_base_offset,
+                        niche_value_u128,
+                        niche_offset_bits,
+                        &untagged_target_component.typ(),
+                        &mut starting_offset,
+                    );
+                    if non_zst_count > 1 {
+                        Expr::union_expr(
+                            overall_t,
+                            untagged_target_component.name(),
+                            untagged_member_expr,
+                            &self.symbol_table,
+                        )
+                    } else {
+                        Expr::struct_expr_with_nondet_fields(
+                            overall_t,
+                            btree_string_map![(
+                                untagged_target_component.name(),
+                                untagged_member_expr
+                            )],
+                            &self.symbol_table,
+                        )
+                    }
+                }
+            }
+        }
+    }
+
+    fn codegen_rvalue_generator(&mut self, operands: &[Operand<'tcx>], ty: Ty<'tcx>) -> Expr {
+        let overall_t = self.codegen_ty(ty);
+        let direct_fields = overall_t.lookup_field("direct_fields", &self.symbol_table).unwrap();
+        let direct_fields_components =
+            direct_fields.typ().lookup_components(&self.symbol_table).unwrap().clone();
+        let mut operands_iter = operands.iter();
+        let direct_fields_expr = Expr::struct_expr_with_nondet_fields(
+            direct_fields.typ(),
+            direct_fields_components
+                .iter()
+                .filter_map(|c| {
+                    if c.is_padding() {
```

Comment:
you might be able to simplify this logic if you iterate over the object layout instead.

Reply:
Implemented as suggested -- it's closer to MIR now rather than trusting information provided by the type codegen layer. I don't know whether this is good or bad.